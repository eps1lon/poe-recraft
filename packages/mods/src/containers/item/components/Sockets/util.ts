import Item from '../../Item';
import { Socket, SocketColor } from './types';
import ValueRange from '../../../../calculator/ValueRange';

/**
 * creates a Socket according to the highest attr requirement
 * if two or more requirements are equal the color is chosen
 * with the priority Green > Red > Blue
 * @param item
 */
export function rollRequirementBiasedStable(item: Item): Socket {
  const { dex, int, str } = item.requirements.list();
  const color_by_attr = [
    { value: dex.value, color: SocketColor.green },
    { value: str.value, color: SocketColor.red },
    { value: int.value, color: SocketColor.blue },
    // descending
  ].sort((a, b) => -ValueRange.cmpMax(a.value, b.value));

  return {
    color: color_by_attr[0].color,
  };
}
