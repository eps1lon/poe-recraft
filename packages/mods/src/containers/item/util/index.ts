export { default as AugmentableValue } from './AugmentableValue';
