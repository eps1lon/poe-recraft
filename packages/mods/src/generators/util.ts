import { GeneratorDetails } from './Generator';

// filter nothing
const allowAll = () => true;

export interface WithSpawnchance {
  spawnchance: number;
}

export const calculateSpawnchance = (
  collection: Array<GeneratorDetails<any>>,
  filter: (details: GeneratorDetails<any>) => boolean = allowAll,
): Array<GeneratorDetails<any> & WithSpawnchance> => {
  const sum_spawnweight = collection.reduce((sum, details) => {
    if (details.spawnweight != null && filter(details)) {
      return sum + details.spawnweight;
    } else {
      return sum;
    }
  }, 0);

  return collection.map(details => {
    let spawnchance = 0;

    if (details.spawnweight != null && filter(details)) {
      spawnchance = details.spawnweight / sum_spawnweight;
    }

    return {
      ...details,
      spawnchance,
    };
  });
};
