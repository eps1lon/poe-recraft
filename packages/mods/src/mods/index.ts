export { default as metaMods } from './meta_mods';
export { default as Mod } from './Mod';
