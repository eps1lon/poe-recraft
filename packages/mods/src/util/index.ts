export { Flags, anySet } from './Flags';
export { default as MetaData } from './MetaData';
