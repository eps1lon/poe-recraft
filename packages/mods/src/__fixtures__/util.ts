import * as fs from 'fs';
import * as path from 'path';

import {
  AtlasNodeProps,
  BaseItemTypeProps,
  CraftingBenchOptionsProps,
  EssenceProps,
  ModProps,
} from '../schema';

import AtlasNode from '../containers/AtlasNode';
import Item from '../containers/item/Item';
import { Essence, MasterBenchOption } from '../generators';
import Mod from '../mods/Mod';
import PropsTable, { TableProps } from '../helpers/PropsTable';
import { Buildable } from '../interfaces/Buildable';

const json_cache: Partial<Tables> = {};

export const createTable = <P extends TableProps, T, A1>(
  file: keyof Tables,
  constructor: Buildable<P, T, A1>,
  arg1: A1,
) => {
  if (json_cache[file] == null) {
    const body = fs.readFileSync(
      path.join(__dirname, `../../data/${file}.json`),
    );

    const json = JSON.parse(body.toString());

    // @ts-ignore
    json_cache[file] = new PropsTable(json, constructor, arg1);
  }

  return json_cache[file];
};

interface Tables {
  atlas: PropsTable<AtlasNodeProps, AtlasNode, undefined>;
  craftingbenchoptions: PropsTable<
    CraftingBenchOptionsProps,
    MasterBenchOption,
    undefined
  >;
  essences: PropsTable<EssenceProps, Essence, ModProps>;
  items: PropsTable<BaseItemTypeProps, Item, undefined>;
  mods: PropsTable<ModProps, Mod, undefined>;
}

export const createTables = () => {
  const tables = {
    atlas: createTable('atlas', AtlasNode, undefined),
    craftingbenchoptions: createTable(
      'craftingbenchoptions',
      MasterBenchOption,
      undefined,
    ),
    items: createTable('items', Item, undefined),
    mods: createTable('mods', Mod, undefined),
  } as Tables;

  // @ts-ignore: its not undefined though?!
  tables.essences = createTable('essences', Essence, tables.mods.all());

  return tables;
};

export const byWorldAreaId = (id: string) => (node: AtlasNodeProps) =>
  node.world_area.id === id;

export const fromWorldAreaId = (
  id: string,
  atlas: PropsTable<AtlasNodeProps, AtlasNode, undefined>,
): AtlasNode => {
  return atlas.from(node => node.world_area.id === id);
};

export const byEssenceName = (name: string) => (essence: EssenceProps) =>
  essence.base_item_type.name === name;
