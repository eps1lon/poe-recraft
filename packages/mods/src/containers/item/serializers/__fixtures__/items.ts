import { createTables } from '../../../../__fixtures__/util';

const { items, mods } = createTables();

export default [
  items
    .fromName('Platinum Kris')
    .rarity.set('magic')
    .addMod(
      mods.fromId('LocalIncreasedPhysicalDamagePercentAndAccuracyRating2'),
    )
    .setProperty('mirrored', true),
  items
    .fromName('Hubris Circlet')
    .rarity.set('rare')
    .asShaperItem()
    .addMod(mods.fromId('IncreasedLife0'))
    .addMod(mods.fromId('LocalIncreasedEnergyShield2'))
    .addMod(mods.fromId('LocalIncreasedEnergyShieldPercentAndStunRecovery3'))
    .addMod(mods.fromId('FireResist1'))
    .addMod(mods.fromId('ReducedLocalAttributeRequirements1'))
    .addMod(mods.fromId('AdditionalMinesPlacedSupportedUber1_'))
    .setProperty('corrupted', true),
];
