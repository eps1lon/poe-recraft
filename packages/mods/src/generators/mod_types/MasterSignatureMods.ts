import { GeneratorDetails } from '../Generator';
import ViewOnlyOrb from '../ViewOnlyOrb';
import Item from '../../containers/item';
import { Mod } from '../../mods';
import { ModProps } from '../../schema';
import { anySet } from '../../util';

/**
 * Generator for all mods that can appear on items sold by Master vendors
 *
 * Remember that items sold from the vendor have a fixed item level and might
 * not be able to hold mods for higher item levels
 */
export default class MasterSignatureMods extends ViewOnlyOrb {
  /**
   * mapping from mod id to master id
   *
   * we need to know what master vendor can spawn the mods because the vendor
   * tells us what itemclasses can appear
   */
  public static readonly MOD_MASTER: { [key: string]: string } = {
    // Leo
    AllDamageMasterVendorItem: 'pvp',
    // Zana
    MapQuantityAddedAsRarityVendorItem: 'str_dex_int',
    // Catarina
    LocalIncreaseSocketedSupportGemLevelIntMasterVendorItem: 'int',
    // Vorici
    LifeLeechSpeedDexIntMasterVendorItem: 'dex_int',
    // Haku
    SocketedGemQualityStrMasterVendorItem: 'str',
    // Tora
    BleedOnHitGainedDexMasterVendorItemUpdated_: 'dex',
    // Vagan
    AlwaysHitsStrDexMasterVendorItem: 'str_dex',
    // Elreon
    ReduceGlobalFlatManaCostStrIntMasterVendor: 'str_int',
  };

  /**
   * mapping from master id to a filter function that returns true if it could've
   * sold the given item
   *
   * TODO required level? Do master sell the best bases e.g. Hubris Circlets
   */
  public static readonly MASTER_ITEMCLASSES: {
    [key: string]: (item: Item) => boolean;
  } = {
    // Leo
    pvp: item => ['Amulet', 'Ring', 'Belt'].includes(item.baseitem.item_class),
    // Zana
    str_dex_int: item => ['Map'].includes(item.baseitem.item_class),
    // Catarina
    int: item =>
      ['Dagger', 'Wand', 'Sceptre'].includes(item.baseitem.item_class) ||
      isSpiritShield(item) ||
      isIntHelmet(item),
    // Vorici
    dex_int: item => ['Gloves', 'Amulet'].includes(item.baseitem.item_class),
    // Haku
    str: item =>
      item.meta_data.isA('AbstractOneHandWeapon') ||
      item.meta_data.isA('AbstractTwoHandWeapon'),
    // Tora
    dex: item => ['Bow'].includes(item.baseitem.item_class),
    // Vagan: 1H/2H melee weapon
    str_dex: item =>
      [
        'One Hand Axe',
        'One Hand Mace',
        'One Hand Sword',
        'Thrusting One Hand Sword',
        'Sceptre',
        'Dagger',
        'Claw',
        'Two Hand Axe',
        'Two Hand Mace',
        'Two Hand Sword',
        'Staff',
      ].includes(item.baseitem.item_class),
    // Elreon
    str_int: item => ['Ring', 'Amulet'].includes(item.baseitem.item_class),
  };

  public static build(mods: ModProps[]): MasterSignatureMods {
    return new MasterSignatureMods(
      mods
        .filter(MasterSignatureMods.isMasterSignature)
        .map(props => new Mod(props)),
    );
  }

  /**
   *
   * @param mod
   * @returns true if the mod can only be found on items bought from masters
   */
  public static isMasterSignature(mod: ModProps) {
    return MasterSignatureMods.MOD_MASTER[mod.id] !== undefined;
  }

  /**
   * @param item
   * @param whitelist
   */
  public modsFor(
    item: Item,
    whitelist: string[] = [],
  ): Array<GeneratorDetails<Mod>> {
    // basically ignore spawnweight
    const whitelist_spawnable = whitelist.concat(
      'no_matching_tags',
      'spawnweight_zero',
    );

    return super
      .modsFor(item, whitelist_spawnable)
      .map(details => {
        const { mod } = details;
        const master = MasterSignatureMods.MOD_MASTER[mod.props.id];
        if (master === undefined) {
          // should not happen unless MOD_MASTER was changed
          throw new Error(`no master found for '${mod.props.id}'`);
        }

        const sellsItem = MasterSignatureMods.MASTER_ITEMCLASSES[master];
        if (sellsItem === undefined) {
          // should not happen unless MASTER_ITEMCLASSES was changed
          throw new Error(`no itemclasses given for '${master}'`);
        }

        return {
          ...details,
          applicable: {
            ...details.applicable,
            not_buyable: !sellsItem(item),
          },
        };
      })
      .filter(({ applicable }) => !anySet(applicable, whitelist_spawnable));
  }
}

function isSpiritShield(item: Item): boolean {
  return (
    item.baseitem.item_class === 'Shield' &&
    item.baseitem.tags.includes('focus')
  );
}

function isIntHelmet(item: Item): boolean {
  return (
    item.baseitem.item_class === 'Helmet' &&
    item.baseitem.tags.includes('int_armour')
  );
}
