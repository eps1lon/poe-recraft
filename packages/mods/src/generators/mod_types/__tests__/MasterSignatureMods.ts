import { createTables } from '../../../__fixtures__/util';
import MasterSignatureMods from '../MasterSignatureMods';
import { Mod } from '../../../mods';

const { items, mods } = createTables();
const generator = MasterSignatureMods.build(mods.all());
const snapshot = ({ mod }: { mod: Mod }) => mod.props.id;

const item_class_samples = {
  helmets: 'Hunter Hood',
  es_helmets: 'Solaris Circlet',
  boots: 'Iron Greaves',
  gloves: 'Iron Gauntlets',
  amulets: 'Amber Amulet',
  rings: 'Iron Ring',
  belts: 'Rustic Sash',
  shields: 'Painted Tower Shield',
  spririt_shields: 'Bone Spirit Shield',
  wands: 'Tornado Wand',
  bows: 'Short Bow',
  daggers: 'Skinning Knife',
  staves: 'Woodful Staff',
  one_hand_melee_weapon: 'Arming Axe',
  two_hand_melee_weapon: 'Bastard Sword',
};

for (const [item_class, item_name] of Object.entries(item_class_samples)) {
  it(`recognizes some mods can only spawn on '${item_class}'`, () => {
    const item = items.fromName(item_name);

    expect(
      generator.modsFor(item, ['domain_full']).map(snapshot),
    ).toMatchSnapshot();
  });
}
