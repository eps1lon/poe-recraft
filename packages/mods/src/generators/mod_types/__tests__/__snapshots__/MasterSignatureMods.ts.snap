// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recognizes some mods can only spawn on 'amulets' 1`] = `
Array [
  "AllDamageMasterVendorItem",
  "ReduceGlobalFlatManaCostStrIntMasterVendor",
  "LifeLeechSpeedDexIntMasterVendorItem",
]
`;

exports[`recognizes some mods can only spawn on 'belts' 1`] = `
Array [
  "AllDamageMasterVendorItem",
]
`;

exports[`recognizes some mods can only spawn on 'boots' 1`] = `Array []`;

exports[`recognizes some mods can only spawn on 'bows' 1`] = `
Array [
  "SocketedGemQualityStrMasterVendorItem",
  "BleedOnHitGainedDexMasterVendorItemUpdated_",
]
`;

exports[`recognizes some mods can only spawn on 'daggers' 1`] = `
Array [
  "LocalIncreaseSocketedSupportGemLevelIntMasterVendorItem",
  "SocketedGemQualityStrMasterVendorItem",
  "AlwaysHitsStrDexMasterVendorItem",
]
`;

exports[`recognizes some mods can only spawn on 'es_helmets' 1`] = `
Array [
  "LocalIncreaseSocketedSupportGemLevelIntMasterVendorItem",
]
`;

exports[`recognizes some mods can only spawn on 'gloves' 1`] = `
Array [
  "LifeLeechSpeedDexIntMasterVendorItem",
]
`;

exports[`recognizes some mods can only spawn on 'helmets' 1`] = `Array []`;

exports[`recognizes some mods can only spawn on 'one_hand_melee_weapon' 1`] = `
Array [
  "SocketedGemQualityStrMasterVendorItem",
  "AlwaysHitsStrDexMasterVendorItem",
]
`;

exports[`recognizes some mods can only spawn on 'rings' 1`] = `
Array [
  "AllDamageMasterVendorItem",
  "ReduceGlobalFlatManaCostStrIntMasterVendor",
]
`;

exports[`recognizes some mods can only spawn on 'shields' 1`] = `Array []`;

exports[`recognizes some mods can only spawn on 'spririt_shields' 1`] = `
Array [
  "LocalIncreaseSocketedSupportGemLevelIntMasterVendorItem",
]
`;

exports[`recognizes some mods can only spawn on 'staves' 1`] = `
Array [
  "SocketedGemQualityStrMasterVendorItem",
  "AlwaysHitsStrDexMasterVendorItem",
]
`;

exports[`recognizes some mods can only spawn on 'two_hand_melee_weapon' 1`] = `
Array [
  "SocketedGemQualityStrMasterVendorItem",
  "AlwaysHitsStrDexMasterVendorItem",
]
`;

exports[`recognizes some mods can only spawn on 'wands' 1`] = `
Array [
  "LocalIncreaseSocketedSupportGemLevelIntMasterVendorItem",
  "SocketedGemQualityStrMasterVendorItem",
]
`;
